include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+    return {&resource_.get(), &destructionEvt_};
+  }
+
+private:
+  _container* set_constructed() noexcept {
+    [[maybe_unused]] auto old =
+        state_.exchange(state_t::CONSTRUCTED, std::memory_order_seq_cst);
+    UNIFEX_ASSERT(old == state_t::DESTRUCT_SPAWNED);
+    return this;
+  }
+
+  UNIFEX_NO_UNIQUE_ADDRESS Scheduler scheduler_;
+  UNIFEX_NO_UNIQUE_ADDRESS unifex::manual_lifetime<Resource> resource_;
+
+  static void destruct_this_resource(_container_base* base) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto& self = static_cast<_container&>(*base);
+    self.resource_.destruct();
+  }
+
+  // TODO potential improvmement to avoid the any_sender_of<> static method:
+  // 1. add operation states to the _container
+  // 2. maybe a union
+  static unifex::any_sender_of<>
+  schedule_destruct(_container_base* base, _container_base* parent) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto& self = static_cast<_container&>(*base);
+    return unifex::sequence(
+        unifex::just_from([&self, parent]() noexcept {
+          [[maybe_unused]] auto old = self.state_.exchange(
+              state_t::DESTRUCT_SPAWNED, std::memory_order_seq_cst);
+          UNIFEX_ASSERT(old == state_t::ALLOCATED);
+          if (parent) {
+            parent->register_child(&self);
+          }
+        }),
+        unifex::unstoppable(unifex::on(self.scheduler_, self.destruct_impl())));
+  }
+
+  static unifex::any_sender_of<>
+  destroy_this_resource(_container_base* base) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto& self = static_cast<_container&>(*base);
+    return unifex::with_query_value(
+        unifex::let_done(async_destroy(self.resource_.get()), unifex::just),
+        unifex::get_scheduler,
+        self.scheduler_);
+  }
+
+  static void deleter(_container_base* base) noexcept {
+    UNIFEX_ASSERT(base != nullptr);
+    auto* self = static_cast<_container*>(base);
+    delete self;
+  }
+};  // namespace unifex
+
+template <typename T>
+struct _type_t {
+  using type = T;
+};
+
+template <
+    typename Resource = void,
+    typename Scheduler,
+    typename ResourceFactory>
+auto async_resource(
+    Scheduler scheduler,
+    async_scope_ref outerScope,
+    ResourceFactory&& factory) noexcept {
+  using factory_result_t =
+      decltype(factory(std::declval<async_scope_ref>(), scheduler));
+  auto result = []() noexcept {
+    if constexpr (unifex::sender<factory_result_t>) {
+      static_assert(
+          !std::is_void_v<Resource>,
+          "Sender returning ResourceFactory must specify Resource");
+      return _type_t<Resource>{};
+    } else {
+      static_assert(
+          unifex::same_as<Resource, void>, "must not specify Resource");
+      return _type_t<factory_result_t>{};
+    }
+  };
+  using resource_t = typename decltype(result())::type;
+  auto* container = new _container<resource_t, Scheduler>{scheduler};
+  auto construct = [&]() noexcept {
```

Comment:
Probably best to leave off the `noexcept`.
```suggestion
  auto construct = [&]() {
```